File: game/card.py
class Card:
    """
    One actual in-game copy of a BrawlerTemplate.
    """

    _instance_counter = 0

    def __init__(
        self,
        template,
        owner_id,
    ):
        Card._instance_counter += 1

        self.instance_id = (
            f"card_{Card._instance_counter}"
        )

        self.template = template
        self.owner_id = owner_id

        self.current_hp = template.max_hp

        self.is_revealed = False
        self.has_acted_once = False
        self.super_unlocked = False
        self.is_defeated = False

        # reserve | active | graveyard
        self.zone = "reserve"

    @property
    def is_alive(self):
        return not self.is_defeated

    def reveal(self):
        self.is_revealed = True

    def mark_acted(self):
        """
        First successful attack unlocks the super.
        """

        if not self.has_acted_once:
            self.has_acted_once = True
            self.super_unlocked = True

    def apply_damage(self, amount):
        self.current_hp = max(
            0,
            self.current_hp - amount,
        )

        if self.current_hp == 0:
            self.is_defeated = True

        return self.current_hp

    def apply_healing(self, amount):
        self.current_hp = min(
            self.template.max_hp,
            self.current_hp + amount,
        )

        return self.current_hp

    def to_public_view(self):
        """
        Information visible to the opponent.
        """

        if not self.is_revealed:
            return {
                "instanceId": self.instance_id,
                "ownerId": self.owner_id,
                "revealed": False,
                "zone": self.zone,
                "alive": self.is_alive,
            }

        return {
            "instanceId": self.instance_id,
            "ownerId": self.owner_id,
            "revealed": True,
            "zone": self.zone,
            "alive": self.is_alive,
            "name": self.template.name,
            "brawlerClass": (
                self.template.brawler_class.value
            ),
            "currentHP": self.current_hp,
            "maxHP": self.template.max_hp,
            "superUnlocked": self.super_unlocked,
        }

    def to_owner_view(self):
        return {
            "instanceId": self.instance_id,
            "ownerId": self.owner_id,
            "revealed": self.is_revealed,
            "zone": self.zone,
            "alive": self.is_alive,
            "name": self.template.name,
            "brawlerClass": (
                self.template.brawler_class.value
            ),
            "currentHP": self.current_hp,
            "maxHP": self.template.max_hp,
            "superUnlocked": self.super_unlocked,
            "hasActedOnce": self.has_acted_once,
        }

File: src/Card.js
/**
 * Card
 * ----
 * A single in-game instance of a BrawlerTemplate. This is what actually
 * lives in a deck/hand/active slot and carries mutable state: current HP,
 * whether it's been revealed, whether its super is unlocked/used, etc.
 */

let instanceCounter = 0;

export class Card {
  /**
   * @param {import("./BrawlerTemplate.js").BrawlerTemplate} template
   * @param {string} ownerId
   */
  constructor(template, ownerId) {
    this.instanceId = `card_${++instanceCounter}`;
    this.template = template;
    this.ownerId = ownerId;

    this.currentHP = template.maxHP;
    this.isRevealed = false;
    this.hasActedOnce = false; // first attack unlocks the super
    this.superUnlocked = false;
    this.isDefeated = false;

    // "reserve" | "active" | "graveyard" — tracked by Deck/Player, mirrored
    // here for convenience when inspecting a single card.
    this.zone = "reserve";
  }

  get isAlive() {
    return !this.isDefeated;
  }

  reveal() {
    this.isRevealed = true;
  }

  /** Call once, right after this card lands its first attack. */
  markActed() {
    if (!this.hasActedOnce) {
      this.hasActedOnce = true;
      this.superUnlocked = true;
    }
  }

  applyDamage(amount) {
    this.currentHP = Math.max(0, this.currentHP - amount);
    if (this.currentHP === 0) {
      this.isDefeated = true;
    }
    return this.currentHP;
  }

  applyHealing(amount) {
    this.currentHP = Math.min(this.template.maxHP, this.currentHP + amount);
    return this.currentHP;
  }

  /**
   * Returns a version of this card's data safe to show to an opponent.
   * Unrevealed cards show only the face-down back; revealed cards show
   * full info.
   */
  toPublicView() {
    if (!this.isRevealed) {
      return {
        instanceId: this.instanceId,
        ownerId: this.ownerId,
        revealed: false,
        zone: this.zone,
        alive: this.isAlive,
        // no name/class/HP leaked — this is the face-down "star design" back
      };
    }
    return {
      instanceId: this.instanceId,
      ownerId: this.ownerId,
      revealed: true,
      zone: this.zone,
      alive: this.isAlive,
      name: this.template.name,
      brawlerClass: this.template.brawlerClass,
      currentHP: this.currentHP,
      maxHP: this.template.maxHP,
      superUnlocked: this.superUnlocked,
    };
  }

  /** Full data, for the owning player's own view or server-side logic. */
  toOwnerView() {
    return {
      ...this.toPublicView(),
      revealed: this.isRevealed, // keep actual reveal flag
      name: this.template.name,
      brawlerClass: this.template.brawlerClass,
      currentHP: this.currentHP,
      maxHP: this.template.maxHP,
      superUnlocked: this.superUnlocked,
      hasActedOnce: this.hasActedOnce,
    };
  }
}
